### PR TITLE
Named exports should always take priority over star exports

### PR DESCRIPTION
**Pull request by arcanis**
This commit fixes https://phabricator.babeljs.io/T2438 by allowing imported symbols to be overriden.

There's still a flag that has to be raised: there might be an ES3 compatibility issue (check the following). However, I feel like an ES3 transform could be made to convert these constructs (ie. defineProperty calls that only use a constexpr getter) into proper ones (ie. defineProperty calls that use values instead of getters and are thus easily polyfillable).

I'm using this code on the development branch of a real-size application, and it works fine.

### CODE REVIEW

**Comment by arcanis** [comment]
This PR isn't yet ready to merge (some identifiers have yet to become string literals), but I'd appreciate some feedback. In the meantime, I'll proceed with updating the test cases.

**Comment by arcanis** [comment]
Here we go :) There's still a flag that has to be raised: there might be an ES3 compatibility issue (check [the following](https://github.com/babel/babel/pull/3468/files#diff-e13ee4cceaabc4864835deb5803c8899R13)). However, I feel like an ES3 transform could be made to convert these constructs (ie. defineProperty calls that only use a static getter) into proper ones (ie. defineProperty calls that use `value` instead of a getter and thus can be easily polyfilled).

**Comment by arcanis** [comment]
I'm now using it in production, and it doesn't work yet. Apparently, the `configurable: true` really needs to be set to true. Is there a reason why it isn't set there?

**Comment by loganfsmyth** [comment]
Rather than relying on import ordering, it seems like it would be ideal if the `export * from` statements explicitly excluded things that were named exports, that way the rest of this logic can stay the same as it currently is, thoughts? e.g.

    export var foo = 4;
    export * from 'foo';

to

    var namedExports = ['foo'];
    exports.foo = undefined;

    Object.keys(foo).forEach(function(key){
        if (key === 'default' || namedExports.indexOf(key) !== -1) return;
        // define prop
    });


    var foo = exports.foo = 4;

**Comment by hzoo** [comment]
That sounds good to me

**Comment by arcanis** [comment]
@loganfsmyth it seems better, indeed. However, I'm not sure I'll have time to do this before a few months. Maybe this PR could be merged as a first iteration, then improved when the need arise?

**Comment by arcanis** [comment]
@loganfsmyth @hzoo any news?

**Comment by arcanis** [comment]
@loganfsmyth @hzoo ? :blush:

**Comment by arcanis** [comment]
Someone ? :confused:

**Comment by arcanis** [comment]
I'm a poor lonesome dev
I'm a long long way from $HOME
And this poor lonesome cowboy
Has got a long long way to merge ♪

Worth noting that even if you're working on this project voluntarily, I'm too (and updating the tests took me a few hours and some sanity that I won't be able to get back). I'd appreciate some kind of answer, even if only to say that this PR is shit and won't be merged.

**Comment by loganfsmyth** [comment]
Hey, you're right we could definitely do better about responding to PRs at least to let you know you're not forgotten, sorry about that.

I still think this approach isn't the best in the long term. If we want to ignore certain re-exported values, that's what we should do, rather than rely on export ordering and property redefinition.

**Comment by arcanis** [comment]
I'm gonna look into it end of the week. Do you know if there's a way to automatically generate the test results ? Fixing the code doesn't really scare me, but updating all those babel outputs to match the new ones was a bit of a traumatic experience :sweat_smile:

**Comment by hzoo** [comment]
Tests are generated for you on babylon tests but not here. We could probably investigate doing the same thing though

**Comment by albertogasparin** [comment]
Nice to see this is moving forward. Well done @arcanis 👍

**Comment by arcanis** [comment]
@loganfsmyth @hzoo Finally got the time to fix this PR. The new implementations adds an `__esExports` symbol (I could have used `__esModule`, but I wasn't sure if changing its value from a boolean to an object would break BC) that contains the list of named exports in the module. This list is then used to prevent `export *` constructs from crashing when colliding with named exports.

These two commits contain the new implementation + the test changes required for the `babel-plugin-transform-es2015-modules-commonjs` plugin. I haven't been able to run the full test on master, it seems there's an error somewhere unrelated to this PR (cf error below).

WDYT?

```
Cannot find module 'babel-helper-plugin-test-runner'
    at Function.Module._resolveFilename (module.js:440:15)
    at Function.Module._load (module.js:388:25)
    at Module.require (module.js:468:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/home/arcanis/Projects/babel/packages/babel-preset-latest/test/index.js:1:63)
    ...
```

**Comment by arcanis** [comment]
Ping @loganfsmyth ?

**Comment by arcanis** [comment]
[This PR is still relevant.](https://babeljs.io/repl/#?babili=false&evaluate=true&lineWrap=false&presets=es2015%2Creact%2Cstage-2&experimental=true&loose=false&spec=true&code=export%20*%20from%20'foo'%3B%0A%0Aexport%20let%20x%20%3D%201%3B)

File: packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js [hunk 1]
```diff
@@ -9,8 +9,13 @@ let buildRequire = template(`
...
 
 let buildExportsModuleDeclaration = template(`
-  Object.defineProperty(exports, "__esModule", {
-    value: true
+  Object.defineProperties(exports, {
+    __esModule: {
+      value: true
+    },
+    __esExports: {
+      value: $0
```
Comment (loganfsmyth) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: Does this need to be stored on `exports`. Couldn't it be a local variable instead?

File: packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js [hunk 2]
```diff
@@ -439,14 +456,16 @@ export default function () {
           }
 
           // add __esModule declaration if this file has any exports
-          if (hasExports && !strict) {
```
Comment (loganfsmyth) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: Not sure this should be removed.

File: packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js [hunk 3]
```diff
@@ -439,14 +456,16 @@ export default function () {
...
             if (this.opts.loose) buildTemplate = buildLooseExportsModuleDeclaration;
 
-            const declar = buildTemplate();
+            const declar = buildTemplate(t.arrayExpression(Object.keys(exportNames).map(function (name) {
+              return t.stringLiteral(name);
+            })));
             declar._blockHoist = 3;
 
-            topNodes.unshift(declar);
+            topNodes = [].concat(declar).concat(topNodes);
```
Comment (loganfsmyth) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: Was this change needed?

Comment (arcanis) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: Yes. Because of the changes in `buildLooseExportsModuleDeclaration`, `declar` might be an array.

File: packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js [hunk 4]
```diff
@@ -439,14 +456,16 @@ export default function () {
           }
 
           // add __esModule declaration if this file has any exports
-          if (hasExports && !strict) {
+          if (hasExports) {
             let buildTemplate = buildExportsModuleDeclaration;
             if (this.opts.loose) buildTemplate = buildLooseExportsModuleDeclaration;
 
-            const declar = buildTemplate();
+            const declar = buildTemplate(t.arrayExpression(Object.keys(exportNames).map(function (name) {
```
Comment (loganfsmyth) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: I'd recommend

```
buildTemplate({
  NAMED_EXPORTS: t.valueToNode(Object.keys(exportNames)),
});
```

to use a name rather than `$0` and `t.valueToNode` to make the array into AST nodes.

File: packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js (hunk 2 above)
Comment (arcanis) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: I haven't been able to understand the purpose of this check (according to git blame, it has been added in the single big commuit between 5.X -> 6.X). It doesn't match the comment above, which is weird.

File: packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js (hunk 1 above)
Comment (arcanis) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: I can change it, but I'll have to find how to properly declare a new variable invisible from the userland code. Is there an easy way to do this?

File: packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js (hunk 2 above)
Comment (loganfsmyth) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: Yeah this argument is poorly named, this is `strict` module transpiling, e.g. without adding a magic `__esModule` export. `strictMode` is the correct flag these days to disable strict mode in modules.

File: packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js (hunk 3 above)
Comment (loganfsmyth) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: Ahh, missed that, good call. Okay, hopefully we can change this back if we stop using `__esExports`

File: packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js (hunk 1 above)
Comment (loganfsmyth) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: You should be able to use:

    // At a point early on so it exists when you call `buildExportAll`
    var namedExportIdentifier = path.scope.generateUidIdentifier('namedExports');

    // At the bottom 
    path.scope.push({id: namedExportIdentifier, init: t.valueToNode(Object.keys(exportNames))});

That's from memory, so if you have issues let me know.

**Comment by arcanis** [comment]
@loganfsmyth I've just force-pushed a new commit. Export names are now kept local to the file, and only included if actually needed (with means that much less tests need to be updated). What do you think?

**Comment by arcanis** [comment]
Ping ? :)

**Comment by arcanis** [comment]
I don't work anymore for the company that made use of the `export *` statement, but I'd still like to merge this PR if possible. Anyone here to review?

**Comment by rickardlindahl** [comment]
I would love to see this, or something like this, being integrated in babel! 

I am currently working with a multisite react application where components are exported in `index.js`-files. For every new "base component" we write we have to add it to all site specific `index.js`-files. It would help us a lot to be able to write something like the following, if we want to override a base component with a site specific one:

```
// site/index.js
export * from 'baseComponents';
export AComponent from './site/AComponent'; // this overrides AComponent that is exported as * above
```

In our current setup we have to do:

```
export {
//  AComponent,
  BComponent,
  ...
  XComponent,
} from 'baseComponent';
export AComponent from './site/AComponent';
```

**Comment by arcanis** [comment]
Ping?

**Comment by arcanis** [comment]
Ping @hzoo? 😃 Would you be interested in a last rebase?

**Comment by jridgewell** [comment]
Please.

**Comment by arcanis** [comment]
I tried to make a quick rebase on the Github interface, let's see if it works!

File: packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js [hunk 5]
```diff
@@ -206,6 +211,9 @@ export default function() {
       },
 
       Program: {
+        enter(path) {
+          this.exportNamesIdentifier = path.scope.generateUidIdentifier("exportNames");
```
Comment (jridgewell) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: Why do this eagerly?

Comment (arcanis) [reply packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: iirc it simply was because I didn't see the point in generating it lazily, since it's only ever used if there's an `export *` statement in the code anyway (cf `needsExportNames`). I guess it could be `null` until we need it (which would also remove the need for `needsExportNames`).

Comment (jridgewell) [reply packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: With this, it "uses" the name so no one else can generate it. Not a major issue.

**Comment by jridgewell** [comment]
Can you fix the failing tests?
